Added insert to list by index

MisterDr opened the pull request:
Added command "LIINSERT" (list index insert) to insert at the list at the specific index.

yoav-steinberg commented:
Was surprised to see this doesn't exist yet. I think we should also consider a corresponding remove by index (range). 
Also I think we should change this to be called `LINSERTAT` and make it possible to pass multiple elements and not just one.
@oranagra WDYT?
This needs rebasing.

MisterDr commented:
- Rebased ( and modified according to the current code )
 - Changed name of the command to LINSERTAT
 - Added corresponding LREMAT command
 
 Although I think that is a nice idea to pass multiple elements I am not sure if that is best approach for this command. Or maybe if LINSERTAT command is changed the LINSERT should be changed too to support multiple values. Maybe someone may throw some suggestion regarding that :)

oranagra commented:
Seems like a good thing to add, but considering 7.0 is in late RC, i'd rather not rush it.
let's add it for 7.2
Thanks.

yoav-steinberg commented on src/commands/linsertat.json:
```diff
@@ -0,0 +1,68 @@
+        "summary": "Insert an element at specific position in a list",
+        "complexity": "O(N) where N is the number of elements to traverse before seeing the value pivot. This means that inserting somewhere on the left end on the list (head) can be considered O(1) and inserting somewhere on the right end (tail) is O(N).",
+        "group": "list",
+        "since": "6.2.6",
```
```suggestion
        "since": "7.2",
```

yoav-steinberg commented on src/commands/linsertat.json:
```diff
@@ -0,0 +1,68 @@
+{
+    "LINSERTAT": {
+        "summary": "Insert an element at specific position in a list",
+        "complexity": "O(N) where N is the number of elements to traverse before seeing the value pivot. This means that inserting somewhere on the left end on the list (head) can be considered O(1) and inserting somewhere on the right end (tail) is O(N).",
```
We should probably support negative values here so it's best if we just say than N is the specified index.

yoav-steinberg commented on src/commands/lremat.json:
```diff
@@ -0,0 +1,51 @@
+        "summary": "Remove elements from a list starting at specified index",
+        "complexity": "O(N+M) where N is the length of the list and M is the number of elements removed.",
+        "group": "list",
+        "since": "6.2.6",
```
```suggestion
        "since": "7.2",
```

yoav-steinberg commented on src/t_list.c:
```diff
@@ -327,6 +327,56 @@ void linsertCommand(client *c) {
+        checkType(c,subject,OBJ_LIST)) return;
+
+    /* Seek pivot from head to tail */
+    iter = listTypeInitIterator(subject,0,LIST_TAIL);
```
I suggest supporting a negative value for running from head.
Also we can probably efficiently do an early error in case the index is out of range, and then let the code assume success in all other cases.

yoav-steinberg commented on src/t_list.c:
```diff
@@ -801,6 +851,56 @@ void lremCommand(client *c) {
+    listTypeIterator *li;
+    if (toremove < 0) {
+        toremove = -toremove;
+        li = listTypeInitIterator(subject,-1,LIST_HEAD);
```
I think a negative index should be used to decide about direction and `toremove` needs to be validated to be larger than 0.

yoav-steinberg replied:
I think a negative index should be used to decide about direction and `toremove` needs to be validated to be larger than 0.

yoav-steinberg commented on src/t_list.c:
```diff
@@ -801,6 +851,56 @@ void lremCommand(client *c) {
 }
 
+/* LREM <key> <count> <index> */
+void lrematCommand(client *c) {
```
I wonder if `listTypeDelRange()` and `listElementsRemoved()` functions can be used to implement this and avoid code duplication?

yoav-steinberg reviewed (commented):
Thanks you. This is excellent.
Lets waif for some comments about multiple values for insert. And hopefully it'll be pushed into 7.2.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redis/redis?pullRequest=3728) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/redis/redis?pullRequest=3728) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: dependabot[bot]<br/>:x: MisterDr<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redis/redis?pullRequest=3728) it.</sub>
